accepted/00000-http-api.md

```diff
@@ -0,0 +1,220 @@
+
+On the HTTP side, Spark allows registration of an endpoint as a "route" per each endpoint method, optionally including a number of named route parameters that construct the route path.
+
+The actual handler methods are designed to use POJOs as parameters. However, Spark expects a specific signature that takes `request` and `response` as parameters for route handlers. A simple wrapper shall unpack the parameters from the request and call the appropriate handler method for each request.
+
+A typical resource-oriented REST API relies on various HTTP verbs to provide contextual information on what type of operation an endpoint is used for. However, the XMLRPC protocol is designed around methods, rather than resources. Therefore, HTTP verbs will be restricted to GET for read-only, and POST for read-write operations regardless of whether they require a payload or not.
+
+Spark provides 3 common ways to attach data to a request. Named route parameters that are part of the base URL, querystring parameters, or the payload itself.
+
+For primitive types, either route parameters or querystring parameters may be used. The desired method shall be specified using a parameter annotation in the handler method signature (explained below).
+
+For complex types, the request payload of type `JSON` must be used. The Spark route wrapper shall unpack the payload using the `GSON` library.
+
+The exposed HTTP routes shall follow the same namespace-method structure with a few extensions:
+
+```
+/rhn/manager/api/namespace/method[/:route/:params][?query=params]
+```
+
+On each request, the Spark route wrapper shall unpack both types of arguments in the request, deserialize the payload, and call the appropriate handler method with unpacked arguments.
+
+To provide a uniform approach to handler registration, Java annotations can be used to specify a method as an endpoint handler. These annotations can be interpreted during initialization to register all the handlers to both frameworks. As a side benefit, all the handler classes in a predetermined package tree can be automatically identified, without the need to add them in a list manually.
+
+Java annotations shall also be used to determine attributes of an endpoint, such as whether it is a read-only method or not.
+
+Below is an example of such an annotation structure:
+
+#### @ApiHandler("namespace") Target=Class
+
+A class marked with this annotation will be registered as the handler of the specified namespace in the XMLRPC framework. The class will further be inspected on initialization to discover any annotated handler methods. Found methods will be registered to Spark as route handlers.
+
+#### @XmlRpcApiEndpoint Target=Method
```

Comment:
> Instead of using these annotations why not have one with a parameter for the list of targets to where we want to expose, with default to all.

That sounds like a good alternative :+1:

> Besides that, I think we should follow the same approach Spring is using: have annotations for each method type: @apiGet, @apiPost, @apiDelete, @apiHead, etc.

Since the handler classes will have the common business logic shared between two protocols, I wanted to keep it independent of a specific protocol as much as possible. Moreover, as mentioned above, we don't plan to use anything other than POST and GET unless we plan a whole new proper REST API, which targets resources rather than methods, because method names already include the semantics, so we don't need different verbs for that.

So considering this, and all the other comments, I guess the best way would be to have just one annotation, with the target protocols as a parameter, something like:
```
@ApiEndpoint(target=[ XMLRPC | HTTP | BOTH(default) ])
```

Reply:
I get your point, and we will not use any other than Get and Post. But my feeling is that we are creating new name to give to something that already has a name.
If we define and use only two verbs, that is a different discussion, but we are basically giving new names, like `ReadOnly`.
If we use the verbs, in future we could easily add support to more if we decide in that way.

Reply:
We need `ReadOnly` in any case, also for the XMLRPC API.

Currently, we use a regex `^(list|get|is|find).*$` against method names to determine if a method is readonly, and restrict access accordingly. IMO this is kinda flaky. I wanted to use this opportunity to change this behavior to be more explicit.

Reply:
Also keep in mind that those annotations could also be used to generate the docs